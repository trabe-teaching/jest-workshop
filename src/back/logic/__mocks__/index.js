export const findUsers = () => ["user1", "user2"];
