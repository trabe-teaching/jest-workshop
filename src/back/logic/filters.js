export const potentialForceUser = ({ rank }) => rank === "potential";

export const forceApprentice = ({ rank }) => rank === "apprentice";

export const darkSideApprentice = user => forceApprentice(user) && user.type === "dark";

export const aliveMaster = ({ rank, status }) => rank === "master" && status === "alive";

export const aliveJediMaster = user => aliveMaster(user) && user.type === "light";

export const aliveSithMaster = user => aliveMaster(user) && user.type === "dark";

export const userMasterIs = id => user => user.master === id;
