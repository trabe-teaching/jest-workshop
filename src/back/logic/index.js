export * from "./banish";
export * from "./betray";
export * from "./find-users";
export * from "./seek-master";
export * from "./turn";
