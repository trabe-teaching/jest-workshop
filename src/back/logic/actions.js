import { potentialForceUser, forceApprentice, aliveJediMaster, darkSideApprentice } from "./filters";

const possibleActions = [
  {
    match: potentialForceUser,
    action: "seek",
  },
  {
    match: forceApprentice,
    action: "turn",
  },
  {
    match: aliveJediMaster,
    action: "banish",
  },
  {
    match: darkSideApprentice,
    action: "betray",
  },
];

export const addActions = user => {
  const actions = possibleActions.reduce((acc, { match, action }) => (match(user) ? [...acc, action] : acc), []);

  return {
    ...user,
    actions,
  };
};
