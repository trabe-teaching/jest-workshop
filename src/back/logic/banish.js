import { userMasterIs } from "./filters";
import { addActions } from "./actions";

export const banish = (datastore, id) => {
  const master = datastore.find(id);
  const apprentices = datastore.findAllBy(userMasterIs(id));

  const forceGhost = { ...master, status: "ghost" };
  const roninApprentices = apprentices.map(apprentice => ({ ...apprentice, master: null }));

  datastore.update(forceGhost);
  roninApprentices.forEach(apprentice => datastore.update(apprentice));

  return [addActions(forceGhost), ...roninApprentices.map(apprentice => addActions(apprentice))];
};
