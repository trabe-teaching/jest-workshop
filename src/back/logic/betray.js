import { addActions } from "./actions";

export const betray = (datastore, id) => {
  const apprentice = datastore.find(id);
  const betrayedMaster = datastore.find(apprentice.master);

  const deadMaster = { ...betrayedMaster, status: "dead" };
  const newMaster = { ...apprentice, rank: "master", master: null };

  datastore.update(deadMaster);
  datastore.update(newMaster);

  return [addActions(deadMaster), addActions(newMaster)];
};
