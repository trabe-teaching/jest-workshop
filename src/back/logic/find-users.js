import { addActions } from "./actions";

export const findUsers = datastore => datastore.all().map(addActions);
