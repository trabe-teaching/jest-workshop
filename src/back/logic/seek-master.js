import { aliveJediMaster } from "./filters";
import { addActions } from "./actions";

export const seekMaster = (datastore, id) => {
  const potential = datastore.find(id);
  const master = datastore.findBy(aliveJediMaster);

  const newApprentice = { ...potential, rank: "apprentice", master: master.id };

  datastore.update(newApprentice);

  return [addActions(newApprentice)];
};
