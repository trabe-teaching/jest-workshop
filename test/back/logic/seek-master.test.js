import { seekMaster } from "back/logic/seek-master";

const seeds = [
  {
    id: 1,
    name: "Luke Skywalker",
    type: "light",
    rank: "master",
    status: "alive",
  },
  {
    id: 4,
    name: "Rey",
    type: "light",
    rank: "potential",
    status: "alive",
  },
];

let data = [...seeds];

expect.addSnapshotSerializer({
  test: val => val && val.seeds && val.changes && val.data,
  print: ({ seeds, changes, data }, serialize) => {
    const buffer = [];
    buffer.push("---------- SEEDS ----------");
    buffer.push(serialize(seeds));
    buffer.push("--------- CHANGES ---------");
    buffer.push(serialize(changes));
    buffer.push("--------- DATA ------------");
    buffer.push(serialize(data));
    return buffer.join("\n");
  },
});

const store = {
  find: id => data[id],
  findBy: criteria => data.filter(criteria)[0],
  update: user => (data = [...data.filter(u => u.id !== user.id), user]),
};

describe("back/logic/seek-master", () => {
  it("converst a potential into an appretince", () => {
    const changes = seekMaster(store, 1);
    expect({ seeds, changes, data }).toMatchSnapshot();
  });
});
