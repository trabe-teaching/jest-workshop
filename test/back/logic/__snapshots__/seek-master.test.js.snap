// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`back/logic/seek-master converst a potential into an appretince 1`] = `
---------- SEEDS ----------
Array [
  Object {
    "id": 1,
    "name": "Luke Skywalker",
    "rank": "master",
    "status": "alive",
    "type": "light",
  },
  Object {
    "id": 4,
    "name": "Rey",
    "rank": "potential",
    "status": "alive",
    "type": "light",
  },
]
--------- CHANGES ---------
Array [
  Object {
    "actions": Array [
      "turn",
    ],
    "id": 4,
    "master": 1,
    "name": "Rey",
    "rank": "apprentice",
    "status": "alive",
    "type": "light",
  },
]
--------- DATA ------------
Array [
  Object {
    "id": 1,
    "name": "Luke Skywalker",
    "rank": "master",
    "status": "alive",
    "type": "light",
  },
  Object {
    "id": 4,
    "master": 1,
    "name": "Rey",
    "rank": "apprentice",
    "status": "alive",
    "type": "light",
  },
]
`;
